Add Comment Image API docs

thomascchen opened the pull request:
This is a first attempt at adding API documentation for Comment Images. 
Never used API Blueprint before 😬  so hope this isn't wildly off the mark.

Closes https://github.com/code-corps/code-corps-api/issues/401

npendery commented on blueprint/api.apib:
```diff
@@ -115,6 +115,18 @@ Exchange credentials for `access_token`.
+## Comment Image [/comment_images]
+
+### Create a Comment Image [POST]
++ Request
```
Just a styling thing here, break out the lines with double spacing

thomascchen replied:
I can make this change, but our spacing seems to be a little inconsistent here. Are there conventions or a style guide for API Blueprint stuff? Or should we identify/create some?

green-arrow replied:
It definitely is inconsistent. In traditional markdown, things that are separated by a line break in the document need a line in between them in the `.md` file. I would vote for the double spacing.

npendery commented on blueprint/api.apib:
```diff
@@ -115,6 +115,18 @@ Exchange credentials for `access_token`.
 
 # Group Comment Images
 
+## Comment Image [/comment_images]
```
This should be `## Comment Image [/comment_images/new]` for the Post

thomascchen replied:
I'm fairly sure that `[/comment_images]` is the correct path for the `create` action. See http://guides.rubyonrails.org/routing.html#crud-verbs-and-actions

green-arrow replied:
If we're following JSON API, then the correct endpoint is `POST /comment_images`

npendery commented on blueprint/api.apib:
```diff
@@ -225,6 +237,28 @@ Exchange credentials for `access_token`.
+## Comment Image (object)
++ filename: `file name` (string)
+
+## Comment Image Relationships Base (object)
```
Shouldn't both the relationships below just be a single object and not an array? ie:
```
+ comment
  + data(Comments Relationship)
+ user
  + data(Users Relationship)
```

thomascchen commented:
Thanks for the comments! 😄  I've fixed the relationship and adjusted the spacing per @npendery and @green-arrow's suggestion, but I'd like to open a separate issue for identifying or creating some set of conventions for our API docs, for things like spacing. I'd prefer to keep this PR focused on accurately describing the API for comment image.

npendery commented:
LGTM! 👍

christopherstyles commented on blueprint/api.apib:
```diff
@@ -115,6 +115,25 @@ Exchange credentials for `access_token`.
+
++ Headers
+
+  Authorization: Basic 5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8
```
This should be `Bearer` instead of `Basic`. Been meaning to open a separate PR to correct it in existing examples. 

Also, do you have any thoughts about potentially replacing the `5e884898da28…` string with maybe a token indicator? So it would look something like: `Authorization: Bearer {access_token}`. I didn’t see a precedent for this in the Blueprint docs, but I think the meaning comes across, and is a little easier to mentally parse than trying to figure out where this string came from.

green-arrow replied:
I agree with replacing it so it becomes `Authorization: Bearer <access_token`. I think that might be better served in its own issue / PR though, just for clarity.

JoshSmith commented:
@thomascchen are you planning more changes here? What are the next steps?

thomascchen commented:
@JoshSmith just wanted to make the change suggested @christopherstyles and @green-arrow had suggested, which I've done now. Also squashed and rebased so this is ready for review again. 👍

npendery commented:
LGTM

joshsmith reviewed (commented):
👍  all looks good.

begedin commented:
@JoshSmith One of us is gonna have to take over here, since this is a whole new thing we're doing here (`index?` and `show?` policies).

# Specs

For this specific case, the only thing that makes sense is
- both endpoints require authentication
- index endpoint sends all records for current user if no filter
- index endpoint allows coalesced ids. We need to fetch records and see if the user has access to each individual record. If any of them are not accessible by user, return a 403, otherwise, return a 200. 
- show endpoint looks fine, we just need tests for it

# Implementation
- for now, we should implement the index endpoint using the `ja_resource` records callback, calling authorize manually for each record in the coalesced id case

I talked to Alan about it, but he mostly reiterated our thoughts on it. There is no good way to do this. My proposal is as following.

# Notes

`stripe-card` is the simpler one. `stripe-subscriptions` are going to be more difficult.

I talked to Alan on api design, but he mainly reiterated my thoughts on it. There is no "correct" way to do this. We need to consider what our endpoint is fetching and decide how it should behave based on that.

joshsmith commented:
@begedin is this ready to review now?

begedin commented:
As discussed, I removed `index` and wrote polices for index, show, delete.

Also wrote controller specs and endpoints. Only thing not working is the create endpoint. Success and validation tests are there and both fail. It's a pretty big thing that isn't working, but it's a step in that direction.

We might need to modify the behavior of the create policy so it doesn't accept a changeset and instead just goes with a hash, which should be relatively simple.

Also, there is a problem with our api case helpers, or more precicely, our json_payload strategy. It relies on the views we defined, so if the view doesn't render some attribute, the payload strategy won't render it either, even though it should, in the case of `create`. 

We might need to build the payload for the create helper manually. I have an idea how to fix the payload strategy, but I'll have to do it at a later time.

joshsmith commented:
@begedin I have now renamed `StripeCard` to `StripePlatformCard` and added a `StripeConnectCard`.

Feel free to apply the last three items on the checklist in the description to `StripePlatformCard`, and then we can review/merge in favor of doing #476 next.

joshsmith commented on test/controllers/stripe_platform_card_controller_test.exs:
```diff
@@ -30,26 +30,42 @@ defmodule CodeCorps.StripePlatformCardControllerTest do
-    test "renders 422 when data is invalid", %{conn: conn, current_user: current_user} do
-      invalid_attrs = %{stripe_token: nil, user: current_user}
-      assert conn |> request_create(invalid_attrs) |> json_response(422)
+      assert conn |> make_create_request(valid_attrs) |> json_response(201)
```
I'm thinking we should do some checking here like we do for the Analytics identify. Was the Stripe API hit? Not sure where the right boundary for tests happens to be but not sure we're hitting it right now.
